keps/sig-apps/961-maxunavailable-for-statefulset/README.md

```diff
@@ -291,80 +686,149 @@ No, the default behavior remains the same.
+   - For new StatefulSets, the behavior will depend on whether the creating API server has the feature gate enabled. If the feature gate is disabled on the API server handling the request, the StatefulSet will default to updating one pod at a time.
 
-If a rolling update is in progress for a StatefulSet, while this feature is being enabled in kube-apiserver
-and kube-controller-manager, the StatefulSet controller can run into corner cases where it will take longer
-for the controller to converge. This will only happen if after enabling the feature, the customer also sets
-maxUnavailable to a number greater than 1,  but the invariants and the logic will ensure that there are never more than
-maxUnavailable pods with the same identity and never more than maxUnavailable being deleted.
+2. During a rolling update of a StatefulSet with the feature enabled:
+   - If the feature gate is disabled in the midst of the rolling update, the behavior of the StatefulSet will shift to the default rolling update mechanism (updating one pod at a time). Existing pods that were terminated as part of the rolling update when the feature was enabled will continue their termination process. However, the creation of new pods will adhere to the default rolling update strategy, potentially leading to a slower update process if `maxUnavailable` was originally set to a value greater than 1.
+
+To summarize, the impact on running StatefulSets during feature enablement and rollback is primarily observed in the rolling update behavior. StatefulSets that begin a rolling update with the feature enabled will utilize the `maxUnavailable` setting until a change in the feature gate status is detected, after which the behavior will revert to the default update strategy.
 
 ###### What specific metrics should inform a rollback?
-TODO when we reach Beta
+
+<!--
+What signals should users be paying attention to when the feature is young
+that might indicate a serious problem?
+-->
+
+Metric Name: statefulset_unavailability_violation
+
+Description: This metric counts the number of times a StatefulSet exceeds its maxUnavailable threshold during a rolling update. This metric increases whenever a StatefulSet is processed and it's observed that spec.replicas - status.readyReplicas > maxUnavailable. The metric is labeled with namespace, statefulset, and a reason label, where reason could be exceededMaxUnavailable. This provides a clear indication of which StatefulSets are not complying with the defined maxUnavailable constraint, allowing for quick identification and remediation.
```

Comment:
IMO we should provide a simple metric like we do in [deployments](https://github.com/kubernetes/kube-state-metrics/blob/122e5e899943eb78eaf3e366733d5dbec6613ac0/internal/store/deployment.go#L221) for max_unavailable instead. 

Having a hard-coded metric like this is not particularly useful because you can change the value of maxUnavailable in the middle of a rollout. And also the feature can potentially change in the future as it depends on other fields. I think it should be enough to provide users with a query that can be used to monitor the statefulsets, without hardcoding the behaviour.

Reply:
I don't necessary agree here, because of two reasons
(1) not everyone is using kube-state-metrics, so depending on that is not ideal
(2) I woud expect changes to maxUnavailable in the middle of the rolllout and things like that a real corner case - and from the perspective a proposed metric actually gives us visibility into whether this feature works.

As I mentioned above, we don't want name/namespace to be on the metric though - to just have a high-level monitoring of the feature.

Reply:
We need to come to a consensus on this one, as this changes what I can use in the KEP to describe troubleshooting, and any other place where metrics are useful. For now I removed and added a comment where I think I need to rewrite based on this decision (will remove the comment on rebases or squash in the end). I particularly think statefulset_unavailability_violation could be useful. But `And also the feature can potentially change in the future as it depends on other fields` seems also like a very good reason to drop it. @wojtek-t what do you think?

Reply:
I´d like to get consensus on this metric discussion. @wojtek-t do you have anything to add with regard to @atiratree latest remarks?

Reply:
I think that @wojtek-t has a point here, relying strictly on kube-state-metrics is not the best. Having a `statefulset_unavailability_violation` metric that is generally available to Kubernetes users sounds better. At least even if we are not able to provide the namespace/name dimensions in Kubernetes, all users will be able to tell that there are unavailability violations from a high level perspective.

The gap that still exists though seems to be the next step for cluster admins. How do they find the list of statefulsets having an issue? That could be done in multiple ways:
1. from the Kubernetes API via object statuses queryable via kubectl like: 
```sh
kubectl get statefulsets -A -o jsonpath='{range .items[?(@.status.conditions[?(@.type=="UnavailabilityViolation")]}{ .metadata.namespace },{ .metadata.name}{"\n"}{end}'
```
2. from the kcm logs
3. from kube-state-metrics' metrics

If we want to keep the method in-tree, I think that adding a new status to the StatefulSet object (if it doesn't already exist) would a decent option.

Reply:
@dgrisonnet what about your opinions on not doing the statefulset_unavailability_violation and keeping it simple with the `kube_statefulset_spec_strategy_rollingupdate_max_unavailable`, similar to what we have for deployments (and the concerns @atiratree has about statefulset_unavailability_violation)?

Reply:
> yes, it is a corner case, but if someone does that, it will start alerting SREs/Admins as they think they can rely on it being accurate

That's true, but in reality alerts go through two different states. Pending and Firing. The former being triggered when the alert is first seen (expression evaluates to true) and the later depend on a user-defined time window meant to avoid false positive and reduce the noise for the cluster administrators.
So, I would expect the cluster-admins to be responsible for creating an alert that would prevent these kind of edge cases to trigger an alert that fires and page them without needing any actions.

> also the app itself may start failing (additional pods become unavailable) during the update for various reasons and that would not be an indication that the feature is not working, but that the app is broken.

It is the responsibility of the app to provide that information to the admins. In any case the admins would be page for either of these reasons:
1. The app is failing
2. An app has an unavailability violation

In both cases the actions they need to take are pretty much the same, although the best would be for them to prioritize the app failure since it would avoid searching for the app having the violation.

Reply:
Adding a condition is a bit more complicated topic and I think would be more preferable as a followup. 

Introducing a new metric into kubernetes sounds okay to me. But I would still prefer if we do not hardcode the behavior into the metric and give more generic metric that users can use according to their needs. We can even supply such query in the proposal.

But what is the general consensus? Should we port all the deployment, statefulset, etc. kube-state-metrics into kubernetes as they might not be available for all users? Some of them have the same importance as this one.

Reply:
>  The app is failing
>    An app has an unavailability violation

These are not the only reasons, what about an application that is just rolling out? The rollout can also depend on different strategies and PodManagementPolicy. This also doesn't take into account future features.

Basically you need to do a full statefulset controller analysis to know your metric is correct. So I would prefer to keep that responsibility in statefulset controller. To either introduce `Available` condition, which is another KEP of itself. Or to make it simpler for now.

Reply:
I need to know which metric would be approved here to finish the KEP sections that would benefit from those metrics (like troubleshooting, informing rollbacks, SLIs, etc if applicable).